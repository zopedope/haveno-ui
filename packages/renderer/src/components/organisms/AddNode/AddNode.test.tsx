// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { describe, expect, it, vi } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { AddNode } from ".";

describe("organisms::AddNode", () => {
  it("renders without exploding", () => {
    const submitSpy = vi.fn();
    const { asFragment, unmount } = render(<AddNode onSubmit={submitSpy} />);
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  it("can render the form without the title", () => {
    const submitSpy = vi.fn();
    const { asFragment, unmount } = render(
      <AddNode showTitle={false} onSubmit={submitSpy} />
    );
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  it("blocks onSubmit on validation failure", async () => {
    const submitSpy = vi.fn();
    const { unmount } = render(
      <AddNode showTitle={false} onSubmit={submitSpy} />
    );
    fireEvent.submit(screen.getByRole("button", { name: "Save" }));
    expect(submitSpy).not.toHaveBeenCalled();
    unmount();
  });

  it("calls onSubmit on successful validation", async () => {
    const submitSpy = vi.fn();
    const user = userEvent.setup();
    const { unmount } = render(
      <AddNode showTitle={false} onSubmit={submitSpy} />
    );
    expect(submitSpy).to.not.toHaveBeenCalled();
    await user.type(
      screen.getByLabelText("Node address"),
      "http://haveno.network"
    );
    await user.type(screen.getByLabelText("Port"), "58080");
    fireEvent.submit(screen.getByRole("button", { name: "Save" }));
    expect(submitSpy).to.toHaveBeenCalledTimes(1);
    unmount();
  });
});
