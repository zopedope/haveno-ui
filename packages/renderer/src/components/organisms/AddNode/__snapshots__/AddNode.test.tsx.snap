// Vitest Snapshot v1

exports[`organisms::AddNode > can render the form without the title 1`] = `
<DocumentFragment>
  <div
    class="mantine-Stack-root mantine-ngu3vw"
  >
    <form>
      <div
        class="mantine-Stack-root mantine-lfk3cq"
      >
        <div
          class="mantine-TextInput-root mantine-1m3pqry"
        >
          <label
            class="mantine-TextInput-label mantine-7802ha"
            for="address"
            id="address-label"
          >
            Node address
            <span
              aria-hidden="true"
              class="mantine-1m203yh mantine-TextInput-required"
            >
               *
            </span>
          </label>
          <div
            class="mantine-TextInput-wrapper mantine-12sbrde"
          >
            <input
              aria-invalid="false"
              aria-label="Node address"
              class="mantine-TextInput-defaultVariant mantine-TextInput-input mantine-yxlzy5"
              id="address"
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="mantine-TextInput-root mantine-1m3pqry"
        >
          <label
            class="mantine-TextInput-label mantine-7802ha"
            for="port"
            id="port-label"
          >
            Port
            <span
              aria-hidden="true"
              class="mantine-1m203yh mantine-TextInput-required"
            >
               *
            </span>
          </label>
          <div
            class="mantine-TextInput-wrapper mantine-12sbrde"
          >
            <input
              aria-invalid="false"
              aria-label="Port"
              class="mantine-TextInput-defaultVariant mantine-TextInput-input mantine-yxlzy5"
              id="port"
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="mantine-TextInput-root mantine-1m3pqry"
        >
          <label
            class="mantine-TextInput-label mantine-7802ha"
            for="user"
            id="user-label"
          >
            Login (optional)
          </label>
          <div
            class="mantine-TextInput-wrapper mantine-12sbrde"
          >
            <input
              aria-invalid="false"
              class="mantine-TextInput-defaultVariant mantine-TextInput-input mantine-yxlzy5"
              id="user"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="mantine-PasswordInput-root mantine-1m3pqry"
        >
          <label
            class="mantine-PasswordInput-label mantine-7802ha"
            for="password"
            id="password-label"
          >
            Password (optional)
          </label>
          <div
            class="mantine-PasswordInput-wrapper mantine-12sbrde"
          >
            <div
              aria-invalid="false"
              class="mantine-PasswordInput-defaultVariant mantine-PasswordInput-input mantine-PasswordInput-input mantine-1fntf85"
            >
              <input
                class="mantine-PasswordInput-innerInput mantine-u0eqyi"
                id="password"
                type="password"
                value=""
              />
            </div>
            <div
              class="mantine-o3oqoy mantine-PasswordInput-rightSection"
            >
              <button
                aria-hidden="true"
                class="mantine-ActionIcon-hover mantine-ActionIcon-root mantine-PasswordInput-visibilityToggle mantine-1vhreye"
                tabindex="-1"
                type="button"
              >
                <svg
                  fill="none"
                  height="15"
                  viewBox="0 0 15 15"
                  width="15"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M7.5 11C4.80285 11 2.52952 9.62184 1.09622 7.50001C2.52952 5.37816 4.80285 4 7.5 4C10.1971 4 12.4705 5.37816 13.9038 7.50001C12.4705 9.62183 10.1971 11 7.5 11ZM7.5 3C4.30786 3 1.65639 4.70638 0.0760002 7.23501C-0.0253338 7.39715 -0.0253334 7.60288 0.0760014 7.76501C1.65639 10.2936 4.30786 12 7.5 12C10.6921 12 13.3436 10.2936 14.924 7.76501C15.0253 7.60288 15.0253 7.39715 14.924 7.23501C13.3436 4.70638 10.6921 3 7.5 3ZM7.5 9.5C8.60457 9.5 9.5 8.60457 9.5 7.5C9.5 6.39543 8.60457 5.5 7.5 5.5C6.39543 5.5 5.5 6.39543 5.5 7.5C5.5 8.60457 6.39543 9.5 7.5 9.5Z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="mantine-1ne3752"
        />
        <div
          class="mantine-Group-root mantine-147cgkf"
        >
          <button
            class="mantine-Button-filled mantine-Button-root mantine-Group-child mantine-1hjgevl"
            type="submit"
          >
            <div
              class="mantine-3xbgk5 mantine-Button-inner"
            >
              <span
                class="mantine-qo1k2 mantine-Button-label"
              >
                Save
              </span>
            </div>
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`organisms::AddNode > renders without exploding 1`] = `
<DocumentFragment>
  <div
    class="mantine-Stack-root mantine-ngu3vw"
  >
    <h3
      class="mantine-Title-root mantine-biockz"
    >
      Add a new node
    </h3>
    <form>
      <div
        class="mantine-Stack-root mantine-lfk3cq"
      >
        <div
          class="mantine-TextInput-root mantine-1m3pqry"
        >
          <label
            class="mantine-TextInput-label mantine-7802ha"
            for="address"
            id="address-label"
          >
            Node address
            <span
              aria-hidden="true"
              class="mantine-1m203yh mantine-TextInput-required"
            >
               *
            </span>
          </label>
          <div
            class="mantine-TextInput-wrapper mantine-12sbrde"
          >
            <input
              aria-invalid="false"
              aria-label="Node address"
              class="mantine-TextInput-defaultVariant mantine-TextInput-input mantine-yxlzy5"
              id="address"
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="mantine-TextInput-root mantine-1m3pqry"
        >
          <label
            class="mantine-TextInput-label mantine-7802ha"
            for="port"
            id="port-label"
          >
            Port
            <span
              aria-hidden="true"
              class="mantine-1m203yh mantine-TextInput-required"
            >
               *
            </span>
          </label>
          <div
            class="mantine-TextInput-wrapper mantine-12sbrde"
          >
            <input
              aria-invalid="false"
              aria-label="Port"
              class="mantine-TextInput-defaultVariant mantine-TextInput-input mantine-yxlzy5"
              id="port"
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="mantine-TextInput-root mantine-1m3pqry"
        >
          <label
            class="mantine-TextInput-label mantine-7802ha"
            for="user"
            id="user-label"
          >
            Login (optional)
          </label>
          <div
            class="mantine-TextInput-wrapper mantine-12sbrde"
          >
            <input
              aria-invalid="false"
              class="mantine-TextInput-defaultVariant mantine-TextInput-input mantine-yxlzy5"
              id="user"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="mantine-PasswordInput-root mantine-1m3pqry"
        >
          <label
            class="mantine-PasswordInput-label mantine-7802ha"
            for="password"
            id="password-label"
          >
            Password (optional)
          </label>
          <div
            class="mantine-PasswordInput-wrapper mantine-12sbrde"
          >
            <div
              aria-invalid="false"
              class="mantine-PasswordInput-defaultVariant mantine-PasswordInput-input mantine-PasswordInput-input mantine-1fntf85"
            >
              <input
                class="mantine-PasswordInput-innerInput mantine-u0eqyi"
                id="password"
                type="password"
                value=""
              />
            </div>
            <div
              class="mantine-o3oqoy mantine-PasswordInput-rightSection"
            >
              <button
                aria-hidden="true"
                class="mantine-ActionIcon-hover mantine-ActionIcon-root mantine-PasswordInput-visibilityToggle mantine-1vhreye"
                tabindex="-1"
                type="button"
              >
                <svg
                  fill="none"
                  height="15"
                  viewBox="0 0 15 15"
                  width="15"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M7.5 11C4.80285 11 2.52952 9.62184 1.09622 7.50001C2.52952 5.37816 4.80285 4 7.5 4C10.1971 4 12.4705 5.37816 13.9038 7.50001C12.4705 9.62183 10.1971 11 7.5 11ZM7.5 3C4.30786 3 1.65639 4.70638 0.0760002 7.23501C-0.0253338 7.39715 -0.0253334 7.60288 0.0760014 7.76501C1.65639 10.2936 4.30786 12 7.5 12C10.6921 12 13.3436 10.2936 14.924 7.76501C15.0253 7.60288 15.0253 7.39715 14.924 7.23501C13.3436 4.70638 10.6921 3 7.5 3ZM7.5 9.5C8.60457 9.5 9.5 8.60457 9.5 7.5C9.5 6.39543 8.60457 5.5 7.5 5.5C6.39543 5.5 5.5 6.39543 5.5 7.5C5.5 8.60457 6.39543 9.5 7.5 9.5Z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="mantine-1ne3752"
        />
        <div
          class="mantine-Group-root mantine-147cgkf"
        >
          <button
            class="mantine-Button-filled mantine-Button-root mantine-Group-child mantine-1hjgevl"
            type="submit"
          >
            <div
              class="mantine-3xbgk5 mantine-Button-inner"
            >
              <span
                class="mantine-qo1k2 mantine-Button-label"
              >
                Save
              </span>
            </div>
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
